### PR TITLE
refactor(storage): use g::c::Options for GrpcClient

### PR DESCRIPTION
**Pull request by coryan**
Change `storage::internal::GrpcClient` and
`storage::internal::HybridClient` to use `google::cloud::Options`. I had
to define exactly where the defaults are applied to an `Options` object,
for now this is in the `*Client::Create()` factory functions, it may get
refined later.

Some additional refactoring in `storage::ClientOptions` and
`storage::internal::CurlClient` to share the initialization of default
options.

Part of the work for #6161

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6209)

### CODE REVIEW

File: google/cloud/storage/internal/grpc_client.cc [hunk 1]
```diff
@@ -48,60 +59,54 @@ bool DirectPathEnabled() {
                         [](absl::string_view v) { return v == "storage"; });
 }
 
-std::string GrpcEndpoint() {
+Options FillWithDefaultsGrpc(Options options) {
+  options = FillWithDefaults(std::make_shared<oauth2::AnonymousCredentials>(),
```
Comment (devjgm) [inline google/cloud/storage/internal/grpc_client.cc:63]: I assume you saw the naming pattern used by Spanner (https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/spanner/internal/defaults.h) and the generator, but following that pattern, this would be named `DefaultOptions()`, with a ` = {}` default argument. I'm  happy to go with whatever name you choose, but just pointing out the difference in case you didn't see that.

File: google/cloud/storage/internal/grpc_client.cc [hunk 2]
```diff
@@ -48,60 +59,54 @@ bool DirectPathEnabled() {
...
-      stub_(
-          google::storage::v1::Storage::NewStub(CreateGrpcChannel(options_))) {}
+std::shared_ptr<GrpcClient> GrpcClient::Create(Options options) {
+  return Create(std::move(options), /*channel_id=*/0);
+}
+
+std::shared_ptr<GrpcClient> GrpcClient::Create(Options options,
+                                               int channel_id) {
+  // Cannot use std::make_shared<> as the constructor is private.
+  return std::shared_ptr<GrpcClient>(
```
Comment (devjgm) [inline google/cloud/storage/internal/grpc_client.cc:101]: Are you planning to call `internal::CheckExpectedOptions()` from somewhere? That should probably be in the first user-facing function that accepts user-provided `options`. I'd imagine that's also the function where you'd call `FillWithDefaults`, which seems like maybe here?

Fine to do in a separate PR.

**Review by devjgm — APPROVED** [review]

File: google/cloud/storage/internal/grpc_client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_client.cc]: Done.  I was not paying attention to be candid.

File: google/cloud/storage/internal/grpc_client.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_client.cc:101]: > Are you planning to call `internal::CheckExpectedOptions()` from somewhere?

Yes.

> Fine to do in a separate PR.

That was my plan, I do not know *exactly* what that first entry point will be, there are a bit too many ways to create `storage::Client` things right now.
